core/src/main/java/org/infinispan/conflict/impl/StateReceiverImpl.java

```diff
@@ -161,26 +161,29 @@ InboundTransferTask createTransferTask(int segmentId, Address source, CacheTopol
          // If an exception is thrown by any of the inboundTransferTasks, then remove all segment results and cancel all tasks
          allSegmentRequests.exceptionally(throwable -> {
             if (trace) log.tracef(throwable, "Exception when processing InboundTransferTask for cache %s", cacheName);
-            cancel(throwable);
-            return null;
+            synchronized (StateReceiverImpl.this) {
+               cancel(throwable);
+               return null;
+            }
          });
 
          future = allSegmentRequests.thenApply(aVoid -> {
             List<Map<Address, CacheEntry<K, V>>> retVal = keyReplicaMap.entrySet().stream()
                   .map(Map.Entry::getValue)
                   .collect(Collectors.toList());
-            clear();
-            return Collections.unmodifiableList(retVal);
+            synchronized (StateReceiverImpl.this) {
+               clear();
+               return Collections.unmodifiableList(retVal);
+            }
          });
          return future;
       }
 
+      // Should only be called when synchronized on StateReceiverImpl.this
```

Comment:
You can use `@GuardedBy("StateReceiverImpl.this")`, although I I'm not sure it this reference is correct and I don't think I ever got IntelliJ to give me a warning about missing synchronization based on those annotations.